# Data Transformation

## Data Transformation
### Step 1 – Understand the Concept (Markdown)

Definition:Data Transformation is the process of converting data into a suitable format or distribution so that it can be used effectively for analysis and Machine Learning.

In simple words:We change the data to make it cleaner, more useful, and easier for ML models to learn.

### Step 2 – Demonstrate the Concept (Markdown)

Common data transformations:

Scaling → Change feature range
Log Transformation → Reduce right skewness
Power Transformation → Make data more normally distributed
Encoding → Convert categorical data into numbers
Binning → Convert continuous values into groups

Example:

Original Salary:
10000, 20000, 50000, 1000000


Log Transformation:
4.0, 4.3, 4.7, 6.0

The large difference between values is reduced.

### Step 3 – Implement the Concept (Python)

In [2]:
import numpy as np

data = [10, 20, 50, 100, 1000]

log_data = np.log1p(data)

log_data

array([2.39789527, 3.04452244, 3.93182563, 4.61512052, 6.90875478])

## Why Transform Data?
### Step 1 – Understand the Concept (Markdown)

Definition: Data is transformed to make it more suitable for analysis and Machine Learning models.

In simple words: We transform data to make it cleaner, balanced, and easier for the model to learn.

Step 2 – Demonstrate the Concept (Markdown)

Main reasons:

Reduce Skewness
→ Makes highly skewed data more balanced.
Handle Large Differences
→ Reduces the effect of very large values.
Improve Model Performance
→ Helps some ML algorithms learn better.
Reduce Outlier Influence
→ Some transformations can reduce the impact of extreme values.
Make Data More Normally Distributed
→ Useful for some statistical and ML techniques.
Improve Relationships
→ Can make relationships between variables easier for models to learn

### Step 3 – Implement the Concept (Python)

In [3]:
import numpy as np

data = [10, 20, 50, 100, 1000]

# Log transformation
transformed_data = np.log1p(data)

print(transformed_data)

[2.39789527 3.04452244 3.93182563 4.61512052 6.90875478]


## Log Transformation
### Step 1 – Understand the Concept (Markdown)

Definition: Log Transformation converts numerical values using a logarithmic function to reduce skewness and the effect of very large values.

In simple words: It compresses large values and makes highly skewed data more balanced.

### Step 2 – Demonstrate the Concept (Markdown)

Example:

Original:       10    100    1000
Log values:    2.30   4.61    6.91

Large differences are reduced.

Useful for:

Right-skewed data
Income
Salary
House prices
Population
Sales

Important:
For zero or positive values, commonly use:

### Step 3 – Implement the Concept (Python)

In [4]:
import numpy as np

data = [10, 100, 1000]

log_data = np.log1p(data)

print(log_data)

[2.39789527 4.61512052 6.90875478]


## Square Root Transformation
### Step 1 – Understand the Concept (Markdown)

Definition:Square Root Transformation converts numerical values using the square root function to reduce skewness and the influence of large values.

In simple words:
It reduces the difference between small and large values and can make right-skewed data more balanced.

### Step 2 – Demonstrate the Concept (Markdown)

Simple example:

Original:     4    16    100
Square Root:  2     4     10

The large value 100 is reduced to 10.

Useful for:

Right-skewed data
Count data
Moderate outliers
Variables such as number of customers, visits, or transactions

In [5]:
import numpy as np

data = [4, 16, 100]

transformed_data = np.sqrt(data)

print(transformed_data)

[ 2.  4. 10.]


## Power Transformation
### Step 1 – Understand the Concept (Markdown)

Definition: Power Transformation changes numerical data using a power function to make the data more normally distributed and reduce skewness.

In simple words:It changes the shape of the data so that it becomes more balanced and suitable for ML models.

### Step 2 – Demonstrate the Concept (Markdown)

Common power transformations:

Box-Cox → works with positive values only
Yeo-Johnson → works with positive, zero, and negative values

Example:

Original data → Highly skewed
                 ↓
          Power Transformation
                 ↓
        More balanced distribution

Useful for:

Reducing skewness
Making data more normally distributed
Improving some ML models

In [6]:
from sklearn.preprocessing import PowerTransformer

data = [[10], [20], [50], [100], [1000]]

transformer = PowerTransformer(method="yeo-johnson")

result = transformer.fit_transform(data)

print(result)

[[-1.3707852 ]
 [-0.70479998]
 [ 0.03698237]
 [ 0.49734043]
 [ 1.54126238]]


## 1Box-Cox Transformation
### Step 1 – Understand the Concept (Markdown)

Definition: Box-Cox Transformation is a power transformation used to reduce skewness and make numerical data closer to a normal distribution.

In simple words: It changes the shape of the data to make it more balanced.

Important:
Box-Cox works only with positive values (> 0).

### Step 2 – Demonstrate the Concept (Markdown)

Example:

Original:
10, 20, 50, 100, 1000
       ↓
   Box-Cox
       ↓
More balanced distribution

Main uses:

Reduce skewness
Make data more normally distributed
Improve some statistical/ML models
### Step 3 – Implement the Concept (Python)


Remember:

Box-Cox → Power Transformation → Positive values only → Reduces skewness.

Difference:

Box-Cox      → Positive values only
Yeo-Johnson  → Positive + Zero + Negative values

In [7]:
from sklearn.preprocessing import PowerTransformer


data = [[10], [20], [50], [100], [1000]]


transformer = PowerTransformer(method="box-cox")


result = transformer.fit_transform(data)


print(result)

[[-1.37926208]
 [-0.69568165]
 [ 0.04133635]
 [ 0.49518363]
 [ 1.53842375]]


## 2Yeo-Johnson Transformation
### Step 1 – Understand the Concept (Markdown)

Definition: Yeo-Johnson Transformation is a power transformation used to reduce skewness and make data more normally distributed.

In simple words:It is similar to Box-Cox, but it can handle positive, zero, and negative values.

### Step 2 – Demonstrate the Concept (Markdown)

Example:

Original:
-100, -10, 0, 10, 100
          ↓
    Yeo-Johnson
          ↓
More balanced distribution

Main advantage:

    Main advantage:

Box-Cox      → Positive values only
Yeo-Johnson  → Positive + Zero + Negative
### Step 3 – Implement the Concept (Python)

       Remember:

Yeo-Johnson → Power Transformation → Handles negative, zero, and positive values → Reduces skewness. 

In [9]:
from sklearn.preprocessing import PowerTransformer

data = [[-100], [-10], [0], [10], [100]]

transformer = PowerTransformer(method="yeo-johnson")

result = transformer.fit_transform(data)

print(result)

[[-1.57329194e+00]
 [-1.57329194e-01]
 [-4.24798640e-15]
 [ 1.57329194e-01]
 [ 1.57329194e+00]]


## Handling Skewed Data
### Step 1 – Understand the Concept (Markdown)

Definition:Handling skewed data means using techniques to reduce the imbalance in a data distribution so it becomes more suitable for analysis and ML.

In simple words: If most values are on one side and a few extreme values are on the other side, the data is skewed.

### Step 2 – Demonstrate the Concept (Markdown)

Types: 
Right Skewed (Positive Skew) → Long tail on the right
Left Skewed (Negative Skew) → Long tail on the left

Common techniques:

Technique	Best Used For
Log Transformation	Right-skewed data
Square Root	Moderate right skew
Box-Cox	Positive data
Yeo-Johnson	Positive, zero & negative data

### Step 3 – Implement the Concept (Python)

In [ ]:
import numpy as np

data = [10, 20, 50, 100, 1000]

# Log transformation
transformed = np.log1p(data)

print(transformed)

In [ ]:
Raw Data
   ↓
Data Cleaning
   ↓
Check Distribution & Outliers
   ↓
Is Data Skewed?
   ├── Yes → Transformation
   │          ├── Log
   │          ├── Square Root
   │          ├── Box-Cox
   │          └── Yeo-Johnson
   │
   └── No / After Transformation
              ↓
        Feature Scaling
              ↓
      ML Model Training

In [ ]:
Why Transform Data?
Log Transformation
Square Root Transformation
Power Transformation
Box-Cox Transformation
Yeo-Johnson Transformation
Handling Skewed Data
Normalization
Standardization
